ENH: Workspace API

richardotis opened the pull request:
- Adds a new `Workspace` object and associated imperative API. This object now underlies the `equilibrium` function, which remains for backwards compatibility. A `Workspace` object can be mutated after creation. The `Workspace.get()` function accepts `ComputableProperty` objects as input and returns arrays. Fixes #154 
- Adds a new `PhaseRecordFactory` object and associated connections to `PhaseRecord`, which enables deferred compilation of symbolically-defined model properties until they are called in the `PhaseRecord` API. The `PhaseRecord` object also gains `prop` and `prop_grad` functions for computing properties and property gradients of arbitrary symbolic attributes of `Model` instances.
- Creates a "Computable Property Framework" (CPF) for calculating properties of the system, individual phases, or some combination of these. It also provides a framework for defining custom properties.
- Adds support for Jansson derivatives (constrained total derivatives aka "dot" derivatives) of thermodynamic properties (fixes #261 )
- Adds support for T0 (tzero) calculation
- Adds support for driving force calculation, including nucleation driving forces, via the `IsolatedPhase` and `DormantPhase` meta-properties. Meta-properties are objects that know how to create new properties (as defined by the CPF). In this case, these meta-properties know how to start sub-calculations . The current way of finding starting points for these calculations is not ideal and prone to failure (basically reuses `CompositionSet` objects from the workspace equilibrium calculation), but it works well in the typical case.
- Adds physical units support via `pint`. All `ComputableProperty` objects have "implementation units" and "display units." Implementation units are what are assumed by the underlying numerical computation (which is done without units, for performance). The display units are what the user gets when `Workspace.get` is called. `ComputableProperty.__getitem__` allows the display units to be changed for individual calls to `Workspace.get`, or when setting `Workspace.conditions`.
- Also regarding physical units, a special unit conversion context is defined to allow conversions between per-mole properties and per-mass properties. There's room to add per-volume here as well, but that may not make it into this PR.
- Adds support for specifying linear combinations and ratios of mole fractions as a condition.
- Adds support for specifying weight fractions as a condition.
- Closes https://github.com/pycalphad/pycalphad/issues/133 through the multi-valued output from `Workspace.get` 
- Closes https://github.com/pycalphad/pycalphad/issues/163 because site fraction conditions can be given now
- Associated tests for these new features.
- `build_phase_records` is removed. Use the `PhaseRecord.prop_*` API in downstream functions instead.

Documentation
- Adds several examples for the Computable Property API, including driving force, T0, and isolated energy surfaces. A demo of export-to-DataFrame is also shown using the new `PandasRenderer`.
- Other examples have been updated to use the new APIs. Examples using the legacy APIs continue to be supported, but have been moved to a new "Advanced Examples" section. These sections will continue to be expanded and reorganized to accommodate our evolving best practices.

* [x] If any dependencies have changed, the changes are reflected in the
  * [x] `setup.py` (runtime requirements)

bocklund commented:
`wks.conditions` needs a `__str__` and/or `__repr__`

sorkins commented:
Hey @richardotis, when I was experimenting with using this branch earlier this year, I made a few minor adjustments due to errors I ran into. It’s possible you already addressed them, or they aren’t relevant anymore, as I haven’t caught up on your recent changes in this PR, but thought I’d send it along for your consideration.

If they don't make sense, I can find some time soon to update my local environment and test again to see if the issues are still there, or if they're already solved and give you more specific examples of where the issues occur.

![image](https://github.com/pycalphad/pycalphad/assets/48386094/a6cf9868-72fa-4b2c-9082-357e563c552f)

sorkins commented:
Here's the link to my fork where I made the changes, which might help more: https://github.com/sorkins/richardotis-pycalphad/tree/dot_derivatives_ss

bocklund commented:
I'm just starting to use this a bit, I'm wondering if I missed if there's a hook to give units for `ModelComputedProperty` that are not derived from the energy. Specifically, I've been looking at `VM`, `V0`, etc

richardotis commented:
> I'm just starting to use this a bit, I'm wondering if I missed if there's a hook to give units for `ModelComputedProperty` that are not derived from the energy. Specifically, I've been looking at `VM`, `V0`, etc

The way it works at the moment is somewhat hacky. You can assign a variable to this module: https://github.com/richardotis/pycalphad/blob/a0541fe69317348d0627ab72440b00c84a3891a0/pycalphad/property_framework/units.py#L22-L36

bocklund commented:
I came across a bug for ternaries when using the cpf for mole fractions specified to be greater than one. Here's some code to reproduce:

```python
@select_database("alnipt.tdb")
def test_jansson_derivative_with_invalid_mass_conditions(load_database):
    """
    CPF values including Jansson derivatives computed for conditions that are invalid should produce NaN.
    """
    # change to load_database() for the actual test
    dbf = load_database()
    wks = Workspace(dbf, ["AL", "NI", "PT"], ["LIQUID"], {v.T: 298.15, v.P: 101325, v.N: 1, v.X("AL"): 0.6, v.X("PT"): 0.6})
    T = wks.get("T")
    assert np.isnan(T)
    GM = wks.get("GM")
    assert np.isnan(GM)
    dGM_dT = wks.get("GM.T")
    assert np.isnan(dGM_dT)
```

```
# The follwing code can be used to test interactively
import numpy as np
from pycalphad import variables as v
from pycalphad.core.workspace import Workspace
from pycalphad import Database
from importlib_resources import files
import pycalphad.tests.databases
dbf = Database(str(files(pycalphad.tests.databases).joinpath("alnipt.tdb")))
wks = Workspace(dbf, ["AL", "NI", "PT"], ["LIQUID"], {v.T: 298.15, v.P: 101325, v.N: 1, v.X("AL"): 0.6, v.X("PT"): 0.6})
T = wks.get("T")  # currently NaN
print(T)
GM = wks.get("GM")  # currently 0
print(GM)
dGM_dT = wks.get("GM.T")  # raises exception
print(dGM_dT)
```

```python
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[23], line 30
     28 GM = wks.get("GM")  # currently 0
     29 print(GM)
---> 30 dGM_dT = wks.get("GM.T")  # raises exception
     31 print(dGM_dT)

File ~/src/pycalphad-workspace-development/pycalphad/pycalphad/core/workspace.py:425, in Workspace.get(self, values_only, *args)
    423         if results.get(arg, None) is None:
    424             results[arg] = np.zeros((arr_size,) + arg.shape)
--> 425         results[arg][local_index, ...] = Q_(arg.compute_property(composition_sets, cur_conds, chemical_potentials),
    426                                             prop_implementation_units).to(prop_display_units, context).magnitude
    427     local_index += 1
    429 for arg in args:

File ~/src/pycalphad-workspace-development/pycalphad/pycalphad/property_framework/computed_property.py:288, in DotDerivativeComputedProperty.compute_property(self, compsets, cur_conds, chemical_potentials)
    286 solver = Solver()
    287 print(compsets)
--> 288 spec = solver.get_system_spec(compsets, cur_conds)
    289 state = spec.get_new_state(compsets)
    290 state.chemical_potentials[:] = chemical_potentials

File ~/src/pycalphad-workspace-development/pycalphad/pycalphad/core/solver.py:54, in Solver.get_system_spec(self, composition_sets, conditions)
     52 from pycalphad.variables import ChemicalPotential, MoleFraction, SiteFraction
     53 compsets = composition_sets
---> 54 state_variables = compsets[0].phase_record.state_variables
     55 nonvacant_elements = compsets[0].phase_record.nonvacant_elements
     56 num_statevars = len(state_variables)

IndexError: list index out of range
```


It doesn't currently seem to affect binaries, i.e. 
```python
X_AL = 2.0
wks = Workspace(dbf, ["AL", "NI"], ["LIQUID"], {v.T: 298.15, v.P: 101325, v.N: 1, v.X("AL"): X_AL})
```

seems to work for any value of `X_AL` and you get results as if `X_AL = 1.0`.

It also breaks for tuple conditions as if they were passed as scalars, for example
```
wks = Workspace(dbf, ["AL", "NI", "PT"], ["LIQUID"], {v.T: 298.15, v.P: 101325, v.N: 1, v.X("AL"): (0, 1, 0.01), v.X("PT"): 0.98})
```
works for the first 2 conditions `[0.00 0.02 0.98]`, `[0.01 0.01 0.98]`, but fails on the third

bocklund commented:
The test assumes that the value of `"GM"` should also be NaN for the invalid conditions, while it's currently zero. NaN is more consistent with the existing pre-workspace behavior, but if there's a good reason for it to be zero, that might be fine

bocklund commented:
Discussed a little out-of-band, we might make the Pint-based units support a little less intrusive initially. Maybe by being opt-in or somehow separating the units from returned quantities so there's fewer UX hiccups with having to call `qty.m`/`qty.magnitude` or have end users see `UnitStrippedWarning` warnings

nury12n commented:
The Jansson/dot derivatives were useful in adapting the Mapping pull request (#517) with the Workspace API; however, there were a few things that I had to workaround.
 - Case 1: binary system where you have one composition set fixed at NP=0 and one free composition set at NP=1  
   - Computing `X(element).T` throws an error. It seems like the `delta_phase_amounts` when you compute the deltas ignores all composition sets with NP=0 and all fixed composition sets, while `v.X.dot_derivative` assumes that the `delta_phase_amounts` will include composition sets with NP>0 and any fixed composition sets (even if NP=0). For this case, this results in an indexing error where `delta_phase_amounts` only has 1 item while `v.X.dot_derivative` will try to index 2 items
   - This also seems to be the case if you have one composition set fixed at NP=0 and several free composition sets with sum(NP)=1
   - The current workaround I doing is to explicitly compute `X(phase, element).T` for all free phases and compute `X(element).T` from there by `dX = x_i*dNP_i + NP_i*dx_i`.
   
 - Case 2: same as case 1, but the free composition set is a stoichiometric phase
   - If computing `X(free phase, element).T`, it will give 0, which is valid since this would be tracing along the phase boundary of the stoichiometric phase. (Following the implementation of `v.X.dot_derivative`, `X(element).T` should also give 0 if it didn't have the error in case 1)
   - If computing `T.X(element)`, this also returns 0 when it should be undefined, since it would be impossible to keep the stoichiometric phase at NP=1 when moving off composition (T would go to positive or negative infinity)
   - This could be a design decision or just how the derivatives end up working out, but for me, it makes it harder to determine if I would be able to step a global condition along a variable `v` since there isn't a way to distinguish whether the derivative with respect to `v` is actually 0 or if it's undefined (which would suggest that I will run into issues when attempting to compute equilibrium).

Here's an example where I take the PT8AL21-LIQUID two phase region and have the LIQUID composition set fixed at NP=0 and the PT8AL21 composition set free at NP=1. 
When computing the deltas for `X('AL')`, every value is 0, so computing `T.X('AL')` will also give 0 instead of undefined.
When computing the deltas for both `X('AL')` and `T`, the delta_phase_amounts has 1 item, which corresponds to PT8AL21. So if I try to compute `X('AL').T`, I get an error where it tries to access `delta_phase_amounts[1]`.
```python
from importlib_resources import files
import numpy as np

from pycalphad import Workspace, Database, variables as v
from pycalphad.core.solver import Solver
import pycalphad.tests.databases

state_vars = {v.N, v.P, v.T}

# Set stoichiometric phase to be free with NP = 1 and liquid to be fixed with NP = 0
def process_cs_list(comp_sets):
    for cs in comp_sets:
        if cs.phase_record.phase_name == 'LIQUID':
            cs.fixed = True
            cs.update(cs.dof[len(state_vars):], 0.0, cs.dof[:len(state_vars)])
        elif cs.phase_record.phase_name == 'PT8AL21':
            cs.fixed = False
            cs.update(cs.dof[len(state_vars):], 1.0, cs.dof[:len(state_vars)])

    # Sort by free phases then fixed phases
    return [cs for cs in comp_sets if not cs.fixed] + [cs for cs in comp_sets if cs.fixed]

def compute_deltas(comp_sets, conds, chem_pots, var):
    # Get deltas (denominator of derivative)
    solver = Solver()
    spec = solver.get_system_spec(comp_sets, conds)
    state = spec.get_new_state(comp_sets)
    state.chemical_potentials[:] = chem_pots
    state.recompute(spec)
    deltas = var.dot_deltas(spec, state)
    return deltas

dbf = Database(str(files(pycalphad.tests.databases).joinpath("alnipt.tdb")))
comps = ["AL", "PT", "VA"]
conds = {v.T: 1250, v.P: 101325, v.X("AL"): 0.8}

# Get composition sets, chemical potentials and conditions (with the fixed/free cs)
wks = Workspace(dbf, comps, conditions=conds)
comp_sets = [cs for _,cs in wks.enumerate_composition_sets()][0]
comp_sets = process_cs_list(comp_sets)
chem_pots = np.squeeze([wks.get(v.MU(e), values_only=True) for e in ["AL", "PT"]])
updated_conds = {var: np.squeeze(var.compute_property(comp_sets, conds, chem_pots)) for var in conds}

# Delta with v.X('AL')
#   Since PT8AL21 is stoichiometric, everything should be undefined here since
#   it would be impossible move the global composition while still having the NP(PT8AL21) = 1
#   However, this gives a list of 0 for everything
x_deltas = compute_deltas(comp_sets, updated_conds, chem_pots, v.X('AL'))
print(f"Deltas for v.X('AL'): {x_deltas}\n")

# T.X('AL') should be undefined, but is 0
dtdx = v.T.dot_derivative(comp_sets, updated_conds, chem_pots, x_deltas)
print(f"T.X('AL') = {dtdx}\n")

# Delta with v.T
#   Since we're tracing along the PT8AL21 phase boundary, which is a stoichiometric phase
#     The delta composition for PT8AL21 should be 0, but it can change for the liquid phase
#     But because the liquid phase is fixed at 0, the overall change in v.X('AL') should also be 0
T_deltas = compute_deltas(comp_sets, updated_conds, chem_pots, v.T)
print(f"Deltas for v.T: {T_deltas}\n")

# Since we trace along PT8AL21, then X('PT8AL21', 'AL').T = X('AL').T = 0
dxdt_phase = v.X('PT8AL21', 'AL').dot_derivative(comp_sets, updated_conds, chem_pots, T_deltas)
print(f"X('PT8AL21', 'AL').T = {dxdt_phase}\n")

# Attempting to compute X('AL').T
# This fails because the delta_phase_amounts seems to skip over all comp sets with NP = 0
#   In this case, the delta_phase_amounts only includes 1 items, which corresponds to PT8AL21
#
# But when computing the derivative of v.X('AL') it seems to assume that only 
# comp sets that were free and NP = 0 were skipped so it assumes that delta_phase_amounts
# will also contain fixed CS with NP = 0 (LIQUID in this case). This causes an indexing 
# error as it assumes delta_phase_amounts has 2 items instead of 1
dxdt = v.X('AL').dot_derivative(comp_sets, updated_conds, chem_pots, T_deltas)
print(f"X('AL').T = {dxdt}\n")
```

```
C:\Users\ury3\OneDrive - LLNL\Documents\pyvenv\pycalphad-workspace\lib\site-packages\numpy\core\fromnumeric.py:45: UnitStrippedWarning: The unit of the quantity is stripped when downcasting to ndarray.
  result = getattr(asarray(obj), method)(*args, **kwds)
Deltas for v.X('AL'): DotDerivativeDeltas(delta_chemical_potentials=array([0., 0.]), delta_statevars=array([0., 0., 0.]), delta_parameters=None, delta_phase_amounts=array([0.]), delta_sitefracs=[array([0., 0.]), array([0., 0.])])

T.X('AL') = 0.0

Deltas for v.T: DotDerivativeDeltas(delta_chemical_potentials=array([-93.96486887,   8.09969835]), delta_statevars=array([0., 0., 1.]), delta_parameters=None, delta_phase_amounts=array([0.]), delta_sitefracs=[array([0., 0.]), array([-0.00034854,  0.00034854])])

X('PT8AL21', 'AL').T = 0.0

Traceback (most recent call last):
  File "c:\Users\ury3\OneDrive - LLNL\Documents\Projects\misc\pycalphad-workspace\mapping\workspace_feedback.py", line 74, in <module>
    dxdt = v.X('AL').dot_derivative(comp_sets, updated_conds, chem_pots, T_deltas)
  File "C:\Users\ury3\OneDrive - LLNL\Documents\Projects\misc\pycalphad-workspace\pycalphad\pycalphad\variables.py", line 421, in dot_derivative
    dot_derivative += deltas.delta_phase_amounts[idx] * func_value
IndexError: index 1 is out of bounds for axis 0 with size 1
```

richardotis commented:
> The Jansson/dot derivatives were useful in adapting the Mapping pull request (#517) with the Workspace API; however, there were a few things that I had to workaround.
> [...]

I'm reproducing this behavior locally using your script and will investigate.

richardotis commented:
@nury12n [70b64f2](https://github.com/pycalphad/pycalphad/pull/432/commits/70b64f2b1cb3da87a524020a630932717293f522) has a fix for the indexing error you encountered. It was a bug in how the `delta_phase_amounts` array was constructed, but it is fixed now for all the `_differential` functions in the minimizer. Please give it a try.

Regarding when some of the deltas should be undefined or zero, I'm looking at that now. For the first case you mentioned, the relevant behavior is happening inside the `fixed_component_differential` function. The claim is that the `delta_statevars` array ought to return undefined for this case (`dot_deltas` is only returning a value from this array, so it's sufficient to analyze this function).

When I analyze the equilibrium matrix that's constructed to compute the Jansson derivative for Case 1, I get
```python
import numpy as np

A = np.array([[ 0.7241,      0.2759,      0.,         -0.        ],  # phase PT8AL21
 [ 0.89772983,  0.10227017,  0.,         -0.        ],               # phase LIQUID
 [ 0.,          0.,          0.,          0.,        ],              # delta X_AL
 [ 0.,          0.,          1.,          0.        ]])             # delta T
rhs = np.array([0., 0., 1., 0.]) # corresponding to delta X_AL being set to 1
```

The equilibrium matrix is rank-deficient for this case, so in principle there should be no unique solution. In the code we use the least-squares solution to improve numerical stability, but the zero-solution we're getting back from lstsq is (I think) spurious. I did just check to confirm that forcing a NaN solution to lstsq for the rank-deficient case will cause your script to return the output you expect for all cases, but it also causes several equilibrium tests to fail, so we may want to handle the calculation in the `_differential` functions differently from the main minimizer code path (which we can easily do).

richardotis commented:
@nury12n I'm confused about your script now. Is it correct that `v.N` is unset and `v.T` is fixed in these calculations? It doesn't make sense to me to put `v.T` in the numerator of a Jansson derivative if it's fixed as a condition.

nury12n commented:
The indexing error seems to work now. I think in the mapping code, I should be able to use the DotDerivativeComputedProperty in place of the workaround that I had.

I realized if I have one of the composition sets fixed, then one of the conditions should be dropped when computing the derivatives. So either `{v.T, v.P, v.N}` or `{v.P, v.N, v.X('AL'}` needs to be defined which should follow how equilibrium is found during mapping.

Here's an updated script that drops one of the variable when computing the deltas. In this case, I drop v.X('AL') when computing deltas for T to compute X('AL').T, and I drop v.T when computing deltas for X('AL') to compute T.X('AL')

```python
from importlib_resources import files
import numpy as np

from pycalphad import Workspace, Database, variables as v
from pycalphad.core.solver import Solver
import pycalphad.tests.databases

state_vars = {v.N, v.P, v.T}

# Set stoichiometric phase to be free with NP = 1 and liquid to be fixed with NP = 0
def process_cs_list(comp_sets, fixed_phase, free_phase):
    for cs in comp_sets:
        if cs.phase_record.phase_name == fixed_phase:
            cs.fixed = True
            cs.update(cs.dof[len(state_vars):], 0.0, cs.dof[:len(state_vars)])
        elif cs.phase_record.phase_name == free_phase:
            cs.fixed = False
            cs.update(cs.dof[len(state_vars):], 1.0, cs.dof[:len(state_vars)])

    # Sort by free phases then fixed phases
    return [cs for cs in comp_sets if not cs.fixed] + [cs for cs in comp_sets if cs.fixed]

def compute_deltas(comp_sets, conds, chem_pots, var, condition_to_drop):
    #Remove condition
    drop_val = conds[condition_to_drop]
    del conds[condition_to_drop]
    print(conds)

    # Get deltas (denominator of derivative)
    solver = Solver()
    spec = solver.get_system_spec(comp_sets, conds)
    state = spec.get_new_state(comp_sets)
    state.chemical_potentials[:] = chem_pots
    state.recompute(spec)
    deltas = var.dot_deltas(spec, state)

    #Restore condition
    conds[condition_to_drop] = drop_val
    return deltas

dbf = Database(str(files(pycalphad.tests.databases).joinpath("alnipt.tdb")))
comps = ["AL", "PT", "VA"]
conds = {v.T: 1250, v.P: 101325, v.X("AL"): 0.8, v.N: 1}

FIXED_PHASE = 'LIQUID'
FREE_PHASE = 'PT8AL21'

# Get composition sets, chemical potentials and conditions (with the fixed/free cs)
wks = Workspace(dbf, comps, conditions=conds)
comp_sets = [cs for _,cs in wks.enumerate_composition_sets()][0]
comp_sets = process_cs_list(comp_sets, FIXED_PHASE, FREE_PHASE)
chem_pots = np.squeeze([wks.get(v.MU(e), values_only=True) for e in ["AL", "PT"]])
updated_conds = {var: np.squeeze(var.compute_property(comp_sets, conds, chem_pots)) for var in conds}

# Delta with v.X('AL')
#   Since PT8AL21 is stoichiometric, everything should be undefined here since
#   it would be impossible move the global composition while still having the NP(PT8AL21) = 1
#   However, this gives a list of 0 for everything
x_deltas = compute_deltas(comp_sets, updated_conds, chem_pots, v.X('AL'), v.T)
print(f"Deltas for v.X('AL'): {x_deltas}\n")

# T.X('AL') should be undefined, but is 0
dtdx = v.T.dot_derivative(comp_sets, updated_conds, chem_pots, x_deltas)
print(f"T.X('AL') = {dtdx}\n")

# Delta with v.T
#   Since we're tracing along the PT8AL21 phase boundary, which is a stoichiometric phase
#     The delta composition for PT8AL21 should be 0, but it can change for the liquid phase
#     But because the liquid phase is fixed at 0, the overall change in v.X('AL') should also be 0
T_deltas = compute_deltas(comp_sets, updated_conds, chem_pots, v.T, v.X('AL'))
print(f"Deltas for v.T: {T_deltas}\n")

# Since we trace along PT8AL21, then X('PT8AL21', 'AL').T = X('AL').T = 0
# If we switch the free phase to LIQUID, then X('LIQUID').T = X('AL').T != 0 (should be -0.0003845)
dxdt_phase = v.X(FREE_PHASE, 'AL').dot_derivative(comp_sets, updated_conds, chem_pots, T_deltas)
print(f"X({FREE_PHASE}, 'AL').T = {dxdt_phase}\n")

dxdt = v.X('AL').dot_derivative(comp_sets, updated_conds, chem_pots, T_deltas)
print(f"X('AL').T = {dxdt}\n")
```

The results for T.X('AL') and X(PT8AL21,'AL').T are the same as before and X('AL').T seems to be working with the indexing changes.
```
C:\Users\ury3\OneDrive - LLNL\Documents\pyvenv\pycalphad-workspace\lib\site-packages\numpy\_core\fromnumeric.py:42: UnitStrippedWarning: The unit of the quantity is stripped when downcasting to ndarray.
  conv = _array_converter(obj)
{P: array(101325.), X_AL: array(0.7241), N: array(1.)}
Deltas for v.X('AL'): DotDerivativeDeltas(delta_chemical_potentials=array([0., 0.]), delta_statevars=array([0., 0., 0.]), delta_parameters=None, delta_phase_amounts=array([0., 0.]), delta_sitefracs=[array([0., 0.]), array([0., 0.])])

T.X('AL') = 0.0

{P: array(101325.), N: array(1.), T: array(1250.)}
Deltas for v.T: DotDerivativeDeltas(delta_chemical_potentials=array([-93.96486887,   8.09969835]), delta_statevars=array([0., 0., 1.]), delta_parameters=None, delta_phase_amounts=array([0., 0.]), delta_sitefracs=[array([0., 0.]), array([-0.00034854,  0.00034854])])

X(PT8AL21, 'AL').T = 0.0

X('AL').T = 0.0
```

If I switch the fixed phase to PT8AL21 and the free phase to LIQUID, this is what I get, which is reasonable with T.X('AL') = 1/X('AL').T
```
C:\Users\ury3\OneDrive - LLNL\Documents\pyvenv\pycalphad-workspace\lib\site-packages\numpy\_core\fromnumeric.py:42: UnitStrippedWarning: The unit of the quantity is stripped when downcasting to ndarray.
  conv = _array_converter(obj)
{P: array(101325.), X_AL: array(0.89772983), N: array(1.)}
Deltas for v.X('AL'): DotDerivativeDeltas(delta_chemical_potentials=array([269596.36434027, -23238.99617509]), delta_statevars=array([    
0.        ,     0.        , -2869.11872041]), delta_parameters=None, delta_phase_amounts=array([-9.01405769e-20,  0.00000000e+00]), delta_sitefracs=[array([ 1., -1.]), array([0., 0.])])

T.X('AL') = -2869.118720407034

{P: array(101325.), N: array(1.), T: array(1250.)}
Deltas for v.T: DotDerivativeDeltas(delta_chemical_potentials=array([-93.96486887,   8.09969835]), delta_statevars=array([0., 0., 1.]), delta_parameters=None, delta_phase_amounts=array([3.14175138e-23, 0.00000000e+00]), delta_sitefracs=[array([-0.00034854,  0.00034854]), array([0., 0.])])

X(LIQUID, 'AL').T = -0.00034853908027135774

X('AL').T = -0.00034853908027135774
```

richardotis commented:
@nury12n Thanks, the new script makes sense. Can you clarify again what's the desired behavior for these cases? If I add some code to compute the residual of the least-squares solution and then throw it away if it's "large," I can cause your latest script to return this:
```
{P: 101325, X_AL: 0.8, N: 1}
Deltas for v.X('AL'): DotDerivativeDeltas(delta_chemical_potentials=array([nan, nan]), delta_statevars=array([ 0.,  0., nan]), delta_parameters=None, delta_phase_amounts=array([ 0., nan]), delta_sitefracs=[array([nan, nan]), array([nan, nan])])

T.X('AL') = nan

{P: 101325, N: 1, T: 1250.0}
Deltas for v.T: DotDerivativeDeltas(delta_chemical_potentials=array([nan, nan]), delta_statevars=array([ 0.,  0., nan]), delta_parameters=None, delta_phase_amounts=array([ 0., nan]), delta_sitefracs=[array([nan, nan]), array([nan, nan])])

X(PT8AL21, 'AL').T = nan

X('AL').T = nan
```
Is this what we want?

nury12n commented:
The behavior that I was thinking of was `T.X('AL') = nan`, `X(PT8AL21, 'AL').T = 0`, `X('AL').T = 0`, where `X(AL).T = 1/T.X(AL)` since the phase boundary for `PT8AL21` is vertical.

For the most part, I don't consider it too big of an issue if `T.X('AL')` is `nan` or `0`, and for mapping, the workaround I used is quite minimal. I think for me, my intuition would say that if I wanted to find the slope of a vertical phase boundary, then `X.T=0` and `T.X=nan`, but it's possible that outputting `nan` values can propagate issues if users were to use `T.X` in other calculations. And maybe the mapping code is a singular use case where the distinction between outputting `nan` vs. `0` can be of use.


For the case in mapping where `T.X` and `X.T` is used:

When mapping a 2-phase region from a node (any 3-phase equilibria in binaries or ternaries), I want to determine which of the two composition sets to fix and which axis direction to step in. If I were to do the Al-Pt system here, it would be something like
1. Find node (Liquid - Pt8Al21 - x)
2. Attempt to start Liquid + Pt8Al21 region
3. Test `T.X('AL')` and `X('AL').T` assuming either Liquid composition set is fixed to NP=0 or Pt8Al21 composition set is fixed (4 cases in total: `T.X('AL'), fixed liquid`, `T.X('AL'), fixed Pt8Al21`, `X('AL').T, fixed liquid`, `X('AL').T, fixed Pt8Al21`)
4. Out of the four derivatives, take the lowest (in absolute value) that is >1, then the phase boundary will start with the associated fixed/free composition set pair and the variable to step in will be the numerator of the derivative. In this case, it would be `T.X('AL'), fixed Pt8Al21` and we step in temperature.

For the Pt8Al21+Liquid case, the absolute value of the derivatives would be `T.X('AL'), fixed Liquid = nan`, `X('AL').T, fixed Liquid = 0`, `T.X('AL'), fixed Pt8Al21 > 1`, and `X('AL').T, fixed Pt8Al21 < 1`. Graphically, `T.X('AL'), fixed Liquid` being undefined would mean that the temperature of the phase boundary would be +inf or -inf when we step in `X('AL')`. Now in this case, `T.X('AL'), fixed Pt8Al21` will be chosen regardless of whether `T.X('AL'), fixed Liquid` is undefined, inf or 0.

But it can be an issue if the two phases are both stoichiometric, where `T.X('AL')` is undefined for either phase being fixed. Currently all 4 derivatives in this case would return 0 and a best direction will not be able to be found from step 4 above. The current workaround I have for this is to sort the derivatives to be `[T.X, X.T]`, so that I can default to choosing `T.X` if no direction is found in step 4. In this case, I want to choose `T.X` and step in temperature, since stepping in composition would make T undefined and result in DLASCLS error messages popping up.

richardotis commented:
[322e310](https://github.com/pycalphad/pycalphad/pull/432/commits/322e310a85be6aad55a5429d7072036a66c2c9fb) fixes the behavior. Above I had messed up the residual calculation by reusing the equilibrium matrix after the call to LAPACK, but LAPACK destroys the original matrix during the calculation. In this commit I've rewritten the call to compute the residual using copies of `A` and `b`, and now we get the expected behavior.

```
{P: 101325, X_AL: 0.8, N: 1}
Deltas for v.X('AL'): DotDerivativeDeltas(delta_chemical_potentials=array([nan, nan]), delta_statevars=array([ 0.,  0., nan]), delta_parameters=None, delta_phase_amounts=array([ 0., nan]), delta_sitefracs=[array([nan, nan]), array([nan, nan])])

T.X('AL') = nan

{P: 101325, N: 1, T: 1250.0}
Deltas for v.T: DotDerivativeDeltas(delta_chemical_potentials=array([-93.96486886,   8.09969833]), delta_statevars=array([0., 0., 1.]), delta_parameters=None, delta_phase_amounts=array([0., 0.]), delta_sitefracs=[array([-0.00034854,  0.00034854]), array([0., 0.])])

X(PT8AL21, 'AL').T = 0.0

X('AL').T = 0.0
```

I'm going to work on writing some tests for this now.

richardotis commented:
Based on feedback, I've changed (in a breaking way) how units work in Workspace. In particular, `Workspace.get` by default no longer returns `pint.Quantity` objects but instead raw numpy arrays, using the `display_units` of the specified properties. Similarly, `Conditions.__getitem__`, `Conditions.items()`, and its other dict-like functions now return numpy arrays in display units, with the ability to switch to implementation units via a `units` kwarg.

I'm interested to know what others think regarding whether `Workspace.get` should also be switched over to the `units` kwarg approach, as well as other feedback on this API change.

On an unrelated note, I've also updated the test suite to catch or filter all warnings, which has enabled us to turn all warnings into errors. I'm working through some py312 deprecation warnings now.

richardotis commented:
Note that as of [8efd521](https://github.com/pycalphad/pycalphad/pull/432/commits/8efd521465f56fff9b3e8ee91bb67351735a1275) all occurences of 'dot derivative' in the API, tests, and docs has been changed over to 'Jansson derivative' to comport with the nomenclature introduced in https://doi.org/10.1016/j.calphad.2024.102705 (cc @cjkunselman18)
In particular, `DotDerivativeComputedProperty` is now `JanssonDerivative`. The string-based syntax (e.g. `HM.T`) remains unchanged and continues to be supported.

bocklund commented:
It seems to me like this also
- Closes https://github.com/pycalphad/pycalphad/issues/133 through the multi-valued output from `Workspace.get` 
- Closes https://github.com/pycalphad/pycalphad/issues/163 because site fraction conditions can be given now

bocklund commented on pycalphad/core/composition_set.pxd:
```diff
@@ -9,4 +9,8 @@ cdef public class CompositionSet(object)[type CompositionSetType, object Composi
     cdef public bint fixed
     cdef readonly double energy
     cdef double[::1] _energy_2d_view
+    cdef FastFunction phase_local_cons_func
+    cdef FastFunction phase_local_cons_jac
```
Are these actually called or used anywhere? I didn't find anywhere that they are accessed.

richardotis replied:
Yes, see my [related comment](https://github.com/pycalphad/pycalphad/pull/432#discussion_r1676952312)

bocklund commented on pycalphad/core/composition_set.pyx:
```diff
@@ -35,18 +36,32 @@ cdef public class CompositionSet(object)[type CompositionSetType, object Composi
         return str(self.__class__.__name__) + "({0}, {1}, NP={2}, GM={3})".format(self.phase_record.phase_name,
                                                                           np.asarray(self.X), self.NP, self.energy)
 
+    cpdef void set_local_conditions(self, dict phase_local_conditions):
+        mod = self.phase_record.phase_record_factory.models[self.phase_record.phase_name]
+        cfuncs = build_phase_local_constraints(mod, self.phase_record.state_variables + self.phase_record.variables,
+                                               phase_local_conditions,
+                                               parameters=self.phase_record.phase_record_factory.param_symbols)
+        self.phase_local_cons_func = FastFunction(cfuncs.internal_cons_func)
+        self.phase_local_cons_jac = FastFunction(cfuncs.internal_cons_jac)
+        self.num_phase_local_conditions = cfuncs.num_internal_cons
+
```
Maybe I'm not understanding, but this function seems to not actually do anything? Related `self.phase_local_cons_func`/`self.phase_local_cons_func` not being read anywhere.

richardotis replied:
`compset.phase_local_cons_jac` is used on minimizer.pyx:65. `PhaseRecord.phase_local_cons_{func,jac}` accepts a `FastFunction` as an argument, and this object is stored on the `CompositionSet`.

bocklund reviewed (changes requested):
`PhaseRecordFactory`, `PhaseRecord` and `CompositionSet` review

richardotis commented:
Another breaking (within this PR) change: As of [fa724d7](https://github.com/pycalphad/pycalphad/pull/432/commits/fa724d7742337b0a7c934b2e1927c445bd40f407) `Workspace.get()` with a single argument now returns the raw float (or array) result, instead of wrapping the value in a list.

richardotis commented on pycalphad/core/phase_rec.pxd:
```diff
@@ -48,20 +80,11 @@ cdef public class PhaseRecord(object)[type PhaseRecordType, object PhaseRecordOb
     cpdef void internal_cons_func(self, double[::1] out, double[::1] dof) nogil
     cpdef void internal_cons_jac(self, double[:,::1] out, double[::1] dof) nogil
     cpdef void internal_cons_hess(self, double[:,:,::1] out, double[::1] dof) nogil
+    cpdef void phase_local_cons_func(self, double[::1] out, double[::1] dof, FastFunction func) nogil
+    cpdef void phase_local_cons_jac(self, double[:, ::1] out, double[::1] dof, FastFunction jac_func) nogil
     cpdef void mass_obj(self, double[::1] out, double[::1] dof, int comp_idx) nogil
     cpdef void mass_obj_2d(self, double[::1] out, double[:, ::1] dof, int comp_idx) nogil
     cpdef void formulamole_obj(self, double[::1] out, double[::1] dof, int comp_idx) nogil
     cpdef void formulamole_grad(self, double[::1] out, double[::1] dof, int comp_idx) nogil
     cpdef void formulamole_hess(self, double[:,::1] out, double[::1] dof, int comp_idx) nogil
-    # Used only to reconstitute if pickled (i.e. via __reduce__)
-    cdef public object ofunc_
-    cdef public object formulaofunc_
-    cdef public object formulagfunc_
-    cdef public object formulahfunc_
-    cdef public object internal_cons_func_
-    cdef public object internal_cons_jac_
-    cdef public object internal_cons_hess_
-    cdef public object massfuncs_
-    cdef public object formulamolefuncs_
-    cdef public object formulamolegradfuncs_
-    cdef public object formulamolehessianfuncs_
+
```
We talked offline and the cache in `FastFunctionFactory` is not persisted through pickle. This could potentially be okay, but we should investigate further.

bocklund commented on pycalphad/core/minimizer.pyx:
```diff
@@ -27,6 +28,17 @@ cdef void lstsq(double *A, int M, int N, double* x, double rcond) nogil:
         for i in range(N):
             x[i] = -1e19
 
+cpdef void lstsq_check_infeasible(double[:,::] A, double[::] b, double[::] out_x):
+    # Note: This function will destroy input matrix A
+    A_copy = np.copy(A)
+    b_copy = np.copy(b)
+    lstsq(&A[0,0], A.shape[0], A.shape[1],
+        &out_x[0], 1e-16)
+    residual = np.sum(np.square(np.dot(A_copy, out_x) - b_copy))
+    if residual > 1e-6:
+        # lstsq solution is spurious; throw it away
+        out_x[:] = np.nan
+
```
It's not a blocker and I'm expecting any action, but it would be nice to have a solution to avoid a copy. This should be infrequent enough because we don't expect to call the `_differential` functions that use this in tight inner loops.

richardotis replied:
If you don't need to check the residual, you can call `lstsq` to avoid the copy. Otherwise, this copy has to occur somewhere.

richardotis commented on pycalphad/codegen/phase_record_factory.py:
```diff
@@ -0,0 +1,78 @@
+        self.nonvacant_elements = sorted([x for x in self.pure_elements if x != 'VA'])
+        self.molar_masses = np.array([dbf.refstates[x]['mass'] for x in self.nonvacant_elements], dtype='float')
+        parameters = parameters if parameters is not None else {}
+        self.models = models
```
Users who want to mutate `models` should mutate `Workspace.models`, which will do the correct cache invalidation of `PhaseRecordFactory` via dependency tracking. `PhaseRecordFactory` is a semi-private (non-toplevel) API and those objects should not be mutated directly by users without doing their own state management.

richardotis commented on pycalphad/codegen/callables.py:
```diff
@@ -1,254 +1,12 @@
-        if verbose:
-            print(name + ' ')
-    return phase_records
+    if output != 'GM':
+        raise ValueError('build_phase_records is deprecated and no longer works when the output keyword '
+                         'is changed from the default. Remove the keyword, and then use the PhaseRecord.prop_* API '
+                         'in downstream functions instead.')
+    return PhaseRecordFactory(dbf, comps, state_variables, models, parameters=parameters)
```
Ok to delete it now if you and @nury12n agree. The information in the error message can be added to the changelog.

richardotis replied:
`build_phase_records` is now removed.

bocklund commented on pycalphad/core/minimizer.pyx:
```diff
@@ -42,13 +54,15 @@ cdef void invert_matrix(double *A, int N, int* ipiv) nogil:
+                               CompositionSet compset, int num_statevars, double[::1] chemical_potentials,
+                               double[::1] phase_dof) nogil:
     "Compute the LHS of Eq. 41, Sundman 2015."
     cdef int comp_idx, i, j, cons_idx, fixed_dof_idx
```
```suggestion
    cdef int comp_idx, i, j
```

bocklund commented on pycalphad/core/minimizer.pyx:
```diff
@@ -79,7 +93,7 @@ cdef void write_row_stable_phase(double[:] out_row, double* out_rhs, int[::1] fr
     free_variable_column_offset += free_stable_compset_indices.shape[0]
     # 1a. This phase row: free state variables
-    for i in range(free_statevar_indices.shape[0]):
+    for i in range(free_statevar_indices.shape[0]):   
```
```suggestion
    for i in range(free_statevar_indices.shape[0]):
```

bocklund commented on pycalphad/core/minimizer.pyx:
```diff
@@ -96,53 +110,55 @@ cdef void write_row_fixed_mole_fraction(double[:] out_row, double* out_rhs, int
                                         double moles_normalization, double[::1] moles_normalization_grad,
-                                        double[::1] phase_amt, int idx):
+                                        double[::1] phase_amt, int idx, double prefactor):
+    if prefactor == 0.0:
+        return
```
Implicitly requires that `out_row` is initialized to zeros. This is currently true, but if we change the behavior, e.g. `equilibrium_matrix = np.empty` instead of `np.zeros`, breaks. 

having the shortcut here is nice, but maybe we just loop over outrow and set to zero explicitly rather than return

richardotis replied:
For linear combination conditions, `write_row_fixed_mole_fraction` is called multiple times in a loop on the same memory. The caller is responsible for initializing whatever `out_row` points to.

richardotis commented on pycalphad/codegen/phase_record_factory.py:
```diff
@@ -0,0 +1,78 @@
+
+    def update_parameters(self, parameters):
+        new_param_symbols, new_param_values = extract_parameters(parameters)
+        if len(new_param_values.shape) > 1:
+            new_param_values = new_param_values[0]
```
It just takes the first set of parameters so it can be initialized with a valid parameter array. The caller is supposed to call `PhaseRecordFactory.update_parameters()` to iterate through a set of parameters in a loop, or use the `PhaseRecord.{obj,prop}_parameters_2d()` API to get a vectorized result based on the set of parameters. In both cases, the caller should be managing that state.

bocklund commented on pycalphad/core/minimizer.pyx:
```diff
@@ -281,48 +300,49 @@ cdef void fill_equilibrium_system(double[::1,:] equilibrium_matrix, double[::1]
-        if self.prescribed_elemental_amounts.shape[0] > 0:
-            self.ALLOWED_MASS_RESIDUAL = min(1e-8, np.min(self.prescribed_elemental_amounts)/10.0)
+        if self.prescribed_mole_fraction_rhs.shape[0] > 0:
+            # With linear combinations of conditions, RHS can now be exactly zero
+            # This means the smallest allowed mass residual needs to be limited to prevent instability
+            self.ALLOWED_MASS_RESIDUAL = max(1e-12, min(1e-8, np.min(np.abs(self.prescribed_mole_fraction_rhs))/10.0))
             # Also adjust mass residual if we are near the edge of composition space
-            self.ALLOWED_MASS_RESIDUAL = min(self.ALLOWED_MASS_RESIDUAL, (1-np.sum(self.prescribed_elemental_amounts))/10.0)
+            self.ALLOWED_MASS_RESIDUAL = min(self.ALLOWED_MASS_RESIDUAL, (1-np.sum(np.abs(self.prescribed_mole_fraction_rhs)))/10.0)
```
Write a test for a linear combinations of conditions where the RHS is negative. Maybe this isn't exactly right, but for example: `v.X("B") - v.X("A") = -0.5`

richardotis replied:
Added `test_lincomb_binary_condition_rhs_negative`

bocklund commented on pycalphad/core/minimizer.pyx:
```diff
@@ -281,48 +300,49 @@ cdef void fill_equilibrium_system(double[::1,:] equilibrium_matrix, double[::1]
-        if self.prescribed_elemental_amounts.shape[0] > 0:
-            self.ALLOWED_MASS_RESIDUAL = min(1e-8, np.min(self.prescribed_elemental_amounts)/10.0)
+        if self.prescribed_mole_fraction_rhs.shape[0] > 0:
+            # With linear combinations of conditions, RHS can now be exactly zero
+            # This means the smallest allowed mass residual needs to be limited to prevent instability
+            self.ALLOWED_MASS_RESIDUAL = max(1e-12, min(1e-8, np.min(np.abs(self.prescribed_mole_fraction_rhs))/10.0))
             # Also adjust mass residual if we are near the edge of composition space
-            self.ALLOWED_MASS_RESIDUAL = min(self.ALLOWED_MASS_RESIDUAL, (1-np.sum(self.prescribed_elemental_amounts))/10.0)
+            self.ALLOWED_MASS_RESIDUAL = min(self.ALLOWED_MASS_RESIDUAL, (1-np.sum(np.abs(self.prescribed_mole_fraction_rhs)))/10.0)
```
While I'm thinking about it, do we have examples for linear combination of conditions in the docs?

richardotis replied:
Added

bocklund reviewed (changes requested):
I haven't reviewed `_differential` code yet, but here's another chunk for the minimizer

bocklund commented on pycalphad/core/solver.py:
```diff
@@ -40,36 +40,81 @@ def get_system_spec(self, composition_sets, conditions):
+            elif isinstance(cond, SiteFraction):
+                # phase-local condition; already handled
+                continue
+            elif str(cond).startswith('W_'):
```
How about 

```suggestion
            elif isinstance(cond, MassFraction) and cond.phase_name is None:
```

richardotis replied:
Fixed

bocklund commented on pycalphad/core/calculate.py:
```diff
@@ -452,44 +452,65 @@ def calculate(dbf, comps, phases, mode=None, output='GM', fake_points=False, bro
+                                  if isinstance(key, v.StateVariable) and hasattr(key, 'phase_name')}
+    str_phase_local_conditions = {str(k): v for k, v in phase_local_conditions.items()}
+    plc_shape = tuple(len(x) for x in phase_local_conditions.values())
+    # TODO: move state variable conditions into conditions dict
```
Is this TODO talking about moving the phase local conditions into the regular conditions dict? something else? is it still relevant to keep in?

richardotis replied:
I eventually want to get rid of `calculate(..., T=300, 1e5)` and only allow passing in a conditions dict. I don't think it's a blocker for this, and I expect I'll end up touching this code again to fully implement fixed-phase conditions.

richardotis commented:
A few notes from offline discussion, still to do:
- Remove `build_phase_records`
- Move the Renderer framework into a separate branch so we can reconcile Workspace and Mapping before coming up with an integrated approach to visualization

bocklund commented:
Capturing some in-person discussion:

We're going to factor the `pycalphad.plot.renderers.Renderer` class out into a separate branch. We want to integrate mapping into `Workspace` so that there is something like `Workspace.map()` to step/map over non-scalar condition dimensions. We want to avoid shipping the `Workspace.plot` feature only to break it once mapping is integrated into `Workspace`. The process is

1. Remove `Renderer` base class and implementations and put them in a separate branch.
2. Merge this PR. The usability of having to have a separate branch for the `Workspace.plot` API doesn't change.
3. Merge the new mapping PR (currently #517 to be superseded by a new PR). This PR will not change any existing PyCalphad code and is purely additive.
4. Write new code (in a new branch) that integrates mapping into `Workspace`, including adding plotting capabilities to the `Workspace` API. That code will supersede the branch created in Step (1).

The Renderers concept is still good and some future renderers could be:
- Matplotlib renderer (currently implemented)
- Pandas renderer (currently implemented but maybe not necessary because it's becoming increasingly simple)
- Other plotting backend renderers (e.g. Bokeh, Vega/Vega-Lite, Plotly)
- ESPEI dataset renderer

bocklund commented on pycalphad/core/workspace.py:
```diff
@@ -0,0 +1,462 @@
+    phase_record_factory: Optional[PhaseRecordFactory] = PRFField(dependsOn=['phases', 'conditions', 'models', 'parameters'])
+    calc_opts: SumType([NoneType, Dict]) = DictField(lambda _: OrderedDict())
+    solver: SolverBase = SolverField(lambda obj: Solver(verbose=obj.verbose), dependsOn=['verbose'])
+    eq: Optional[LightDataset] = EquilibriumCalculationField(dependsOn=['phase_record_factory', 'calc_opts', 'solver'])
```
Not sure where it lives (in `Workspace` or in `TypedField` or both), but it would be nice to document that `eq` is set by a callback from `TypedField`

richardotis replied:
Comment added in `Workspace`

bocklund commented on pycalphad/core/hyperplane.pyx:
```diff
@@ -56,16 +56,113 @@ cdef int argmax(double* a, int a_shape) nogil:
+    free(constraint_matrix)
+    free(constraint_rhs)
+
+#@cython.boundscheck(False)
```
commented code

richardotis replied:
Fixed

bocklund commented on pycalphad/core/hyperplane.pyx:
```diff
@@ -56,16 +56,113 @@ cdef int argmax(double* a, int a_shape) nogil:
+    cdef double* target_point = <double*>malloc(compositions.shape[1] * sizeof(double))
+    cdef int* int_tmp = <int*>malloc(simplex_size * sizeof(int))
+    cdef size_t[::1] free_chempot_indices = np.array(list(set(range(compositions.shape[1])) - set(fixed_chempot_indices)), dtype=np.uintp)
+    #print('free_chempot_indices', np.asarray(free_chempot_indices))
```
```suggestion
```

bocklund commented on pycalphad/core/hyperplane.pyx:
```diff
@@ -56,16 +56,113 @@ cdef int argmax(double* a, int a_shape) nogil:
+    intersecting_point(compositions, fixed_chempot_indices, trial_simplex,
+                       fixed_lincomb_molefrac_coefs, fixed_lincomb_molefrac_rhs,
+                       <double[:compositions.shape[1]]>target_point)
+    #print('target_point ', np.asarray(<double[:compositions.shape[1]]>target_point))
```
```suggestion
```

bocklund commented on pycalphad/core/hyperplane.pyx:
```diff
@@ -56,16 +56,113 @@ cdef int argmax(double* a, int a_shape) nogil:
+    #print('target_point ', np.asarray(<double[:compositions.shape[1]]>target_point))
+    # Fill coordinate matrix
+    for j in range(simplex_size):
+        # compositions[trial_simplex[i], :]
```
```suggestion
```

bocklund commented on pycalphad/core/hyperplane.pyx:
```diff
@@ -56,16 +56,113 @@ cdef int argmax(double* a, int a_shape) nogil:
+            constraint_rhs[i] = fixed_lincomb_molefrac_rhs[i]
+        constraint_matrix[fixed_lincomb_molefrac_rhs.shape[0] + j*compositions.shape[1]] = plane_coefs[j]
+        constraint_rhs[fixed_lincomb_molefrac_rhs.shape[0]] = 1
+    #with gil:
+    #    print('constraint_matrix ', np.asarray(<double[:fixed_lincomb_molefrac_rhs.shape[0] + 1, :compositions.shape[1]]>constraint_matrix))
+    #    print('constraint_rhs ', np.asarray(<double[:fixed_lincomb_molefrac_rhs.shape[0] + 1]>constraint_rhs))
+    #raise ValueError('stop')
```
```suggestion
```

bocklund commented on pycalphad/core/hyperplane.pyx:
```diff
@@ -56,16 +56,113 @@ cdef int argmax(double* a, int a_shape) nogil:
+    cdef double* constraint_matrix = <double*>malloc((fixed_lincomb_molefrac_rhs.shape[0] + 1) * compositions.shape[1] * sizeof(double))
+    cdef double* constraint_rhs = <double*>malloc((fixed_lincomb_molefrac_rhs.shape[0] + 1) * sizeof(double))
+    out_intersecting_point[:] = 0
+    # out_intersecting_point memory is reused for plane_coefs, to save an allocation
```
```suggestion
```

bocklund commented on pycalphad/core/hyperplane.pyx:
```diff
@@ -56,16 +56,113 @@ cdef int argmax(double* a, int a_shape) nogil:
+    # out_intersecting_point memory is reused for plane_coefs, to save an allocation
+    cdef double[::1] plane_coefs = out_intersecting_point
+    hyperplane_coefficients(compositions, fixed_chempot_indices, trial_simplex, plane_coefs)
+    #with gil:
+    #    print('plane_coefs ', np.asarray(plane_coefs))
```
```suggestion
```

bocklund commented on pycalphad/core/hyperplane.pyx:
```diff
@@ -56,16 +56,113 @@ cdef int argmax(double* a, int a_shape) nogil:
+        for i in range(compositions.shape[1]):
+            out_intersecting_point[i] = compositions[trial_simplex[0], i]
+        return
+    #with gil:
+    #    print('trial_simplex ', np.asarray(trial_simplex))
```
```suggestion
```

richardotis replied:
Fixed

bocklund commented on pycalphad/core/hyperplane.pyx:
```diff
@@ -170,39 +259,24 @@ cpdef double hyperplane(double[:,::1] compositions,
-            for i in range(simplex_size):
-                for j in range(simplex_size):
-                    f_contig_trial[i + j*simplex_size] = f_trial_matrix[i + j*simplex_size + trial_idx*simplex_size*simplex_size]
+            #print('trial simplex ', np.asarray(<int[:simplex_size]>&trial_simplices[trial_idx*simplex_size]))
```
```suggestion
```

richardotis replied:
Fixed

bocklund commented on pycalphad/core/hyperplane.pyx:
```diff
@@ -170,39 +259,24 @@ cpdef double hyperplane(double[:,::1] compositions,
+                              fixed_lincomb_molefrac_coefs, fixed_lincomb_molefrac_rhs, &fractions[trial_idx*simplex_size])
             smallest_fractions[trial_idx] = _min(&fractions[trial_idx*simplex_size], simplex_size)
-
+        #print('smallest_fractions ', np.asarray(<double[:simplex_size]>&smallest_fractions[0]))
```
```suggestion
```

richardotis replied:
Fixed

bocklund commented on pycalphad/core/hyperplane.pyx:
```diff
@@ -170,39 +259,24 @@ cpdef double hyperplane(double[:,::1] compositions,
+        #print('smallest_fractions ', np.asarray(<double[:simplex_size]>&smallest_fractions[0]))
         # Choose simplex with the largest smallest-fraction
         saved_trial = argmax(smallest_fractions, simplex_size)
+        #print('saved_trial', saved_trial)
```
```suggestion
```

richardotis replied:
Fixed

bocklund commented on pycalphad/core/hyperplane.pyx:
```diff
@@ -170,39 +259,24 @@ cpdef double hyperplane(double[:,::1] compositions,
         # Should be exactly one candidate simplex
         for i in range(simplex_size):
             candidate_simplex[i] = trial_simplices[saved_trial*simplex_size + i]
+        #print('candidate_simplex ', np.asarray(<int[:simplex_size]>&candidate_simplex[0]))
```
```suggestion
```

bocklund commented on pycalphad/core/lower_convex_hull.py:
```diff
@@ -88,46 +59,115 @@ def lower_convex_hull(global_grid, state_variables, result_array):
     while not it.finished:
-        indep_idx = []
+        primary_index = it.multi_index
+        # grid_index is constructed at every iteration, based on state variables (independent potentials)
```
```suggestion
```

bocklund commented on pycalphad/core/utils.py:
```diff
@@ -93,20 +94,24 @@ def unpack_condition(tup):
     elif type(phases) is str:
         active_phases = [phases]
+    else:
+        raise ValueError('Cannot unpack phases into recognizable input')
```
```suggestion
        raise ValueError(f'Cannot unpack phases into recognizable input. Got {phases} of type {type(phases)}')
```

bocklund commented on pycalphad/property_framework/units.py:
```diff
@@ -0,0 +1,79 @@
+    else:
+        return qt
+
+energy_implementation_units = GM_implementation_units = 'J / mol'
+energy_display_units = GM_display_units = 'J / mol'
+energy_display_name = GM_display_name = 'Gibbs Energy'
+G_implementation_units = 'J'
+G_display_units = 'J'
+G_display_name = 'Gibbs Energy'
```
These are duplicated in `Model`. Also repeating the nit that the `energy_implementation_units` should be `G` not `GM`

richardotis replied:
Addressed in the [other comment](https://github.com/pycalphad/pycalphad/pull/432#discussion_r1677169198)

bocklund commented on pycalphad/property_framework/tzero.py:
```diff
@@ -0,0 +1,100 @@
+    return None
+
+class T0(object):
+    "T0: (GM(ONE) - GM(TWO))**2 = 0"
```
```suggestion
    "T0: temperature where the energy of two phases are equal, GM(ONE) = GM(TWO)"
```

richardotis commented on pycalphad/plot/ternary.py:
```diff
@@ -45,7 +45,7 @@ def ternplot(dbf, comps, phases, conds, x=None, y=None, eq_kwargs=None, **plot_k
     """
     eq_kwargs = eq_kwargs if eq_kwargs is not None else dict()
     indep_comps = [key for key, value in conds.items() if isinstance(key, v.MoleFraction) and len(np.atleast_1d(value)) > 1]
-    indep_pots = [key for key, value in conds.items() if (type(key) is v.StateVariable) and len(np.atleast_1d(value)) > 1]
+    indep_pots = [key for key, value in conds.items() if isinstance(key, v.IndependentPotential) and len(np.atleast_1d(value)) > 1]
```
Per offline discussion, we're going to defer support for this

richardotis commented on pycalphad/variables.py:
```diff
@@ -370,25 +617,96 @@ class ChemicalPotential(StateVariable):
+
+temperature = T = TemperatureType()
+pressure = P = PressureType()
+moles = N = SystemMolesType()
```
I don't think it makes eventually adding `v.N("A")` difficult. The `v.N` object currently has essentially no API. We could do some tricks with `__call__` or something to make `v.N` and `v.N("A")` work at the same time.

richardotis commented on setup.py:
```diff
@@ -34,7 +34,7 @@ def read(fname):
     description='CALPHAD tools for designing thermodynamic models, calculating phase diagrams and investigating phase equilibria.',
     # Do NOT include pycalphad._dev here. It is for local development and should not be distributed.
-    packages=['pycalphad', 'pycalphad.codegen', 'pycalphad.core', 'pycalphad.io', 'pycalphad.plot', 'pycalphad.plot.binary', 'pycalphad.tests', 'pycalphad.tests.databases', 'pycalphad.models'],
+    packages=['pycalphad', 'pycalphad.codegen', 'pycalphad.core', 'pycalphad.io', 'pycalphad.plot', 'pycalphad.plot.binary', 'pycalphad.property_framework', 'pycalphad.tests', 'pycalphad.tests.databases', 'pycalphad.models'],
```
For one, we don't want it to pick up `pycalphad._dev`

richardotis commented on pycalphad/variables.py:
```diff
@@ -370,25 +617,96 @@ class ChemicalPotential(StateVariable):
         return 'MU_%s' % self.species.name
 
-temperature = T = StateVariable('T')
-entropy = S = StateVariable('S')
-pressure = P = StateVariable('P')
-volume = V = StateVariable('V')
```
I don't want there to be confusion about how to refer to properties. For entropy, the correct thing to do is `ModelComputedProperty('SM')` (or just use `as_property('SM')`. Volume is going to need some work to make things like `GM.P` and `ModelComputedProperty('VM')` consistent, for both calculation and condition purposes.

richardotis commented on setup.py:
```diff
@@ -34,7 +34,7 @@ def read(fname):
     description='CALPHAD tools for designing thermodynamic models, calculating phase diagrams and investigating phase equilibria.',
     # Do NOT include pycalphad._dev here. It is for local development and should not be distributed.
-    packages=['pycalphad', 'pycalphad.codegen', 'pycalphad.core', 'pycalphad.io', 'pycalphad.plot', 'pycalphad.plot.binary', 'pycalphad.tests', 'pycalphad.tests.databases', 'pycalphad.models'],
+    packages=['pycalphad', 'pycalphad.codegen', 'pycalphad.core', 'pycalphad.io', 'pycalphad.plot', 'pycalphad.plot.binary', 'pycalphad.property_framework', 'pycalphad.tests', 'pycalphad.tests.databases', 'pycalphad.models'],
```
See https://github.com/pycalphad/pycalphad/pull/432#discussion_r1677160485

richardotis commented on setup.py:
```diff
@@ -61,13 +61,16 @@ def read(fname):
         'setuptools_scm[toml]>=6.0',
         'symengine>=0.9.2',  # python-symengine on conda-forge
         'tinydb>=3.8',
+        'typing_extensions', # drop when pycalphad drops support for Python<3.8
```
Removed

richardotis commented on setup.py:
```diff
@@ -61,13 +61,16 @@ def read(fname):
         'setuptools_scm[toml]>=6.0',
         'symengine>=0.9.2',  # python-symengine on conda-forge
         'tinydb>=3.8',
+        'typing_extensions', # drop when pycalphad drops support for Python<3.8
```
Removed

richardotis commented on pycalphad/core/workspace.py:
```diff
@@ -0,0 +1,462 @@
+import warnings
+from collections import OrderedDict, Counter, defaultdict
+from collections.abc import Mapping
```
Removed `collections.abc` import

richardotis commented on pycalphad/core/workspace.py:
```diff
@@ -0,0 +1,462 @@
+from pycalphad.property_framework import ComputableProperty, as_property
+from pycalphad.property_framework.units import unit_conversion_context, ureg, as_quantity, Q_
+from runtype import isa
+from runtype.pytypes import Dict, List, Sequence, SumType, Mapping, NoneType
```
We need the `runtype.pytypes` versions of these types to make all the magic casting / type coercion work. We don't construct instances of these types, so the concrete types (e.g. `dict`, `list`) that users may be checking in downstream code should be unaffected by our choice to use `runtype.pytypes` internally.

richardotis commented on pycalphad/core/workspace.py:
```diff
@@ -0,0 +1,462 @@
+
+
+
+def _adjust_conditions(conds) -> 'OrderedDict[StateVariable, List[float]]':
```
Fixed

richardotis commented on pycalphad/core/workspace.py:
```diff
@@ -0,0 +1,462 @@
+
+
+
+def _adjust_conditions(conds) -> 'OrderedDict[StateVariable, List[float]]':
```
The usage of low-level APIs in `pycalphad/plot/binary/map.py` complicates removing `_adjust_conditions` yet, but I do think we can do it once we've removed the legacy mapping code.

richardotis commented on pycalphad/model.py:
```diff
@@ -430,9 +431,12 @@ def degree_of_ordering(self):
     #pylint: disable=C0103
     # These are standard abbreviations from Thermo-Calc for these quantities
     energy = GM = property(lambda self: self.ast)
+    energy_implementation_units = GM_implementation_units = 'J / mol'
+    energy_display_units = GM_display_units = 'kJ / mol'
     formulaenergy = G = property(lambda self: self.ast * self._site_ratio_normalization)
```
These specific lines will be deleted because it's duplicative with what's in `pycalphad/property_framework/units.py`, but your point still stands.

`energy_{implementation,display}_units` needs to have consistent units with `Model.energy`, which is equal to `Model.GM`.

Eventually we can use the units support to move between formula units and per-mole-atom units and deprecate the `M` suffix defined directly on `Model`, but we're not all the way there yet.

richardotis commented on pycalphad/core/workspace.py:
```diff
@@ -0,0 +1,462 @@
+    def cast_from(cls, key: SumType([str, StateVariable])) -> "ConditionKey":
+        return as_property(key)
+
+class TypedField:
+    def __init__(self, default_factory=None, dependsOn=None):
+        self.default_factory = default_factory
+        self.dependsOn = dependsOn
+
```
Fixed

richardotis commented on pycalphad/core/workspace.py:
```diff
@@ -0,0 +1,462 @@
+    def __init__(self, default_factory=None, dependsOn=None):
+        self.default_factory = default_factory
+        self.dependsOn = dependsOn
+
```
Docstring added

richardotis commented on pycalphad/core/workspace.py:
```diff
@@ -0,0 +1,462 @@
+            try:
+                value = self.type.cast_from(value)
+            except TypeError as e:
+                raise e
```
Fixed

richardotis commented on pycalphad/property_framework/computed_property.py:
```diff
@@ -0,0 +1,312 @@
+        return hash(str(self))
+
+    @property
+    def shape(self):
+        return tuple()
```
Added

richardotis commented on pycalphad/property_framework/computed_property.py:
```diff
@@ -0,0 +1,312 @@
+        return tuple()
+
+    @property
+    def multiplicity(self):
```
Added

richardotis commented on pycalphad/core/equilibrium.py:
```diff
@@ -164,18 +36,10 @@ def equilibrium(dbf, comps, phases, conditions, output=None, model=None,
-        If None, return a Dask graph of the computation instead of actually doing it.
     parameters : dict, optional
         Maps SymEngine Symbol to numbers, for overriding the values of parameters in the Database.
     solver : pycalphad.core.solver.SolverBase
```
Removed

richardotis commented on pycalphad/core/constraints.py:
```diff
@@ -19,3 +19,28 @@ def build_constraints(mod, variables, parameters=None):
+    import pycalphad.variables as v
+    phase_local_constraints = []
+    for key, value in phase_local_conditions.items():
+        # Should each phase-local condition key have a `.as_equation(model)` function?
```
Removed

richardotis commented on pycalphad/core/conditions.py:
```diff
@@ -0,0 +1,182 @@
+
+        if idx is None:
+            # Condition is not yet specified
+            # TODO: Check number of degrees of freedom
```
Deferred (TODO removed)

richardotis commented on pycalphad/core/equilibrium.py:
```diff
@@ -190,117 +54,39 @@ def equilibrium(dbf, comps, phases, conditions, output=None, model=None,
+        output = set()
+    elif (not isinstance(output, Iterable)) or isinstance(output, str):
+        output = [output]
+    wks = Workspace(database=dbf, components=comps, phases=phases, conditions=conditions, models=model, parameters=parameters,
+                    verbose=verbose, calc_opts=calc_opts, solver=solver, phase_record_factory=phase_records)
```
`make_prf` is covered by the tests and is used by `default_factory`, so it needs to be a separate method.

richardotis commented on pycalphad/core/conditions.py:
```diff
@@ -0,0 +1,182 @@
+            # "Zero" composition is a common pattern. Do not warn for that case.
+            if np.any(np.logical_and(np.asarray(vals) < self.minimum_composition, np.asarray(vals) > 0)):
+                warnings.warn(
+                    f"Some specified compositions are below the minimum allowed composition of {self.minimum_composition}.")
```
1.0-epsilon is the de facto maximum allowed composition today, but I don't think it needs its own warning here because `v.X(A) = 1 - [tiny value]` is a much less common pattern than `v.X(A) = [tiny value]`.

richardotis commented on pycalphad/model.py:
```diff
@@ -1355,7 +1359,7 @@ def shift_reference_state(self, reference_states, dbe, contrib_mods=None, output
 
```
Deprecation warning added.

richardotis commented on pycalphad/core/constraints.py:
```diff
@@ -19,3 +19,28 @@ def build_constraints(mod, variables, parameters=None):
                            internal_cons_hess=internal_cons_hess,
                            num_internal_cons=len(internal_constraints))
+
+def build_phase_local_constraints(mod, variables, phase_local_conditions, parameters=None):
```
I like having `build_phase_local_constraints` grouped with `build_constraints` in the same module. They are very similar concerns and relate to different parts of the codebase (`CompositionSet` and `Model`, respectively). If it were refactored, they would then be separate.

As discussed previously, we use all of the `FastFunction` objects created by `CompositionSet.set_local_conditions` when we call `PhaseRecord.phase_local_cons_{func,jac}`.

richardotis replied:
You're right that we don't use the Hessian currently, but it's almost always zero anyway.

richardotis commented on pycalphad/core/workspace.py:
```diff
@@ -0,0 +1,462 @@
+    def copy(self):
+        return copy(self)
+
+    @property
+    def plot(self):
+        return self.renderer
+
```
Moved to [`richardotis:renderer_framework`](https://github.com/richardotis/pycalphad/tree/renderer_framework)

richardotis commented on pycalphad/plot/renderers.py:
```diff
@@ -0,0 +1,94 @@
+from pycalphad.property_framework import ComputableProperty, as_property
+from pycalphad.property_framework.units import ureg
+from typing import Tuple
```
Moved to [`richardotis:renderer_framework`](https://github.com/richardotis/pycalphad/tree/renderer_framework)

bocklund commented on pycalphad/property_framework/tzero.py:
```diff
@@ -0,0 +1,100 @@
+        s = self.solver
+        initial_conditions = cur_conds
+
+        # T0: (G(BCC) - G(HCP))**2 = 0
+        # G(BCC)**2 - 2*G(BCC)*G(HCP) + G(BCP)**2
+        # grad = 2*G(BCC)*G'(FCC) - 2*(G'(BCC)*G(HCP) + G'(HCP)*G(BCC)) + 2*G(HCP)*G'(HCP)
```
```suggestion
```
